libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayHistorySession.kt

```diff
@@ -0,0 +1,184 @@
+ * in order to improve or demo an experience. But for the sake of example, this is how you would
+ * replay an experience that just happened.
+ * ```
+ * mapboxNavigation.historyRecorder.stopRecording { historyFile ->
+ *   historyFile?.let {
+ *     replayHistorySession.setHistoryFile(historyFile)
+ *     replayHistorySession.onAttached(mapboxNavigation)
+ *   }
+ * }
+ * ```
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+class ReplayHistorySession : MapboxNavigationObserver {
+
+    private val optionsFlow = MutableStateFlow(ReplayHistorySessionOptions.Builder().build())
+    private var mapboxHistoryReader: MapboxHistoryReader? = null
+    private var mapboxNavigation: MapboxNavigation? = null
+    private var lastHistoryEvent: ReplayEventBase? = null
+    private var coroutineScope: CoroutineScope? = null
+
+    private val replayEventsObserver = ReplayEventsObserver { events ->
+        if (optionsFlow.value.enableSetRoute) {
+            events.filterIsInstance<ReplaySetNavigationRoute>().forEach(::setRoute)
+        }
+        if (isLastEventPlayed(events)) {
+            pushMorePoints()
+        }
+    }
+
+    /**
+     * Signals that the [mapboxNavigation] instance is ready for use.
+     *
+     * @param mapboxNavigation
+     */
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
+            .also { this.coroutineScope = it }
+        this.mapboxNavigation = mapboxNavigation
+        this.lastHistoryEvent = null
+        mapboxNavigation.startReplayTripSession()
+        mapboxNavigation.mapboxReplayer.stop()
+        mapboxNavigation.mapboxReplayer.registerObserver(replayEventsObserver)
+        observeStateFlow(mapboxNavigation).launchIn(coroutineScope)
+    }
+
+    /**
+     * Signals that the [mapboxNavigation] instance is being detached.
+     *
+     * @param mapboxNavigation
+     */
+    override fun onDetached(mapboxNavigation: MapboxNavigation) {
+        coroutineScope?.cancel()
+        coroutineScope = null
+        this.mapboxNavigation = null
+        mapboxNavigation.mapboxReplayer.unregisterObserver(replayEventsObserver)
+        mapboxNavigation.mapboxReplayer.stop()
+        mapboxNavigation.mapboxReplayer.clearEvents()
+    }
+
+    /**
+     * Allows you to get or observe the [ReplayHistorySessionOptions].
+     */
+    fun getOptions(): StateFlow<ReplayHistorySessionOptions> = optionsFlow.asStateFlow()
+
+    /**
+     * Update the [ReplayHistorySessionOptions].
+     */
+    fun setOptions(options: ReplayHistorySessionOptions) {
+        optionsFlow.value = options
+    }
+
+    /**
+     * Change the history file to replay, if changed after [onAttached], the previous trip will
+     * be reset and the new file will start.
+     */
+    fun setHistoryFile(absolutePath: String) {
+        optionsFlow.value = optionsFlow.value.toBuilder().filePath(absolutePath).build()
```

Comment:
This is a good use case for `update`:
```suggestion
        optionsFlow.update { it.toBuilder().filePath(absolutePath).build() }
```